src/TextareaPrompt.php

```diff
@@ -0,0 +1,207 @@
+
+        $this->cursorPosition = $fullLines->sum() + $newColumn;
+    }
+
+    /**
+     * Handle the down keypress.
+     */
+    protected function handleDownKey(): void
+    {
+        $lines = collect($this->lines());
+
+        // Line length + 1 for the newline character
+        $lineLengths = $lines->map(fn ($line, $index) => mb_strlen($line) + ($index === $lines->count() - 1 ? 0 : 1));
+
+        $currentLineIndex = $this->currentLineIndex();
+
+        if ($currentLineIndex === $lines->count() - 1) {
+            // They're already at the last line, jump them to the last position
+            $this->cursorPosition = mb_strlen($lines->implode(PHP_EOL));
+
+            return;
+        }
+
+        // Lines up to and including the current line
+        $currentLines = $lineLengths->slice(0, $currentLineIndex + 1);
+
+        $currentColumn = $currentLines->last() - ($currentLines->sum() - $this->cursorPosition);
+
+        $destinationLineLength = ($lineLengths->get($currentLineIndex + 1) ?? $currentLines->last()) - 1;
+
+        $newColumn = min(max(0, $destinationLineLength), $currentColumn);
+
+        $this->cursorPosition = $currentLines->sum() + $newColumn;
+    }
+
+    /**
+     * The currently visible options.
+     *
+     * @return array<int, string>
+     */
+    public function visible(): array
+    {
+        $this->adjustVisibleWindow();
+
+        $withCursor = $this->valueWithCursor(10_000);
+
+        return array_slice(explode(PHP_EOL, $withCursor), $this->firstVisible, $this->scroll, preserve_keys: true);
+    }
+
+    protected function adjustVisibleWindow(): void
+    {
+        if (count($this->lines()) < $this->scroll) {
+            return;
+        }
+
+        $currentLineIndex = $this->currentLineIndex();
+
+        if ($this->firstVisible + $this->scroll <= $currentLineIndex) {
+            $this->firstVisible++;
+        }
+
+        if ($currentLineIndex === $this->firstVisible - 1) {
+            $this->firstVisible = max(0, $this->firstVisible - 1);
+        }
+
+        // Make sure there are always the scroll amount visible
+        if ($this->firstVisible + $this->scroll > count($this->lines())) {
+            $this->firstVisible = count($this->lines()) - $this->scroll;
+        }
+    }
+
+    /**
+     * Get the index of the current line that the cursor is on.
+     */
+    protected function currentLineIndex(): int
+    {
+        $totalLineLength = 0;
+
+        return (int) collect($this->lines())->search(function ($line) use (&$totalLineLength) {
+            $totalLineLength += mb_strlen($line) + 1;
+
+            return $totalLineLength > $this->cursorPosition;
+        }) ?: 0;
+    }
+
+    /**
+     * Get the formatted lines of the current value.
+     *
+     * @return array<int, string>
+     */
+    public function lines(): array
+    {
+        // TODO: Figure out the real number here, this comes from the renderer?
+        $value = wordwrap($this->value(), 59, PHP_EOL, true);
```

Comment:
Any ideas on this one? I'm thinking it would be something similar to what we do with the `scroll` property to make sure it will fit on the screen.

Reply:
I took a cue from how you were handling the spinner interval and set a public property on the `Textarea` class from the renderer once the proper width was calculated. I think this solves it if you are good with that, tried it out at a bunch of different sizes and it looked good to me.